src/Symfony/Component/HttpFoundation/Session/Storage/NativeSessionStorage.php

```diff
@@ -377,37 +395,29 @@ public function setOptions(array $options)
      * @param \SessionHandlerInterface|null $saveHandler
      *
      * @throws \InvalidArgumentException
+     *
+     * @final since version 3.4
      */
     public function setSaveHandler($saveHandler = null)
     {
         if (!$saveHandler instanceof AbstractProxy &&
             !$saveHandler instanceof \SessionHandlerInterface &&
             null !== $saveHandler) {
-            throw new \InvalidArgumentException('Must be instance of AbstractProxy; implement \SessionHandlerInterface; or be null.');
-        }
-
-        if ($saveHandler instanceof AbstractProxy) {
-            @trigger_error(
-                'Using session save handlers that are instances of AbstractProxy is deprecated since version 3.4 and will be removed in 4.0.',
-                E_USER_DEPRECATED
-            );
+            throw new \InvalidArgumentException('Must implement \SessionHandlerInterface; or be null.');
         }
 
         if (headers_sent($file, $line)) {
             throw new \RuntimeException(sprintf('Failed to set the session handler because headers have already been sent by "%s" at line %d.', $file, $line));
         }
 
-        // Wrap $saveHandler in proxy and prevent double wrapping of proxy
-        if (!$saveHandler instanceof AbstractProxy && $saveHandler instanceof \SessionHandlerInterface) {
+        if (null === $saveHandler) {
+            $saveHandler = new SessionHandlerWrapper(new \SessionHandler());
+        } elseif (!$saveHandler instanceof \SessionHandlerInterface) {
```

Comment:
doesn't this break when session handler is `AbstractProxy`. AbstractProxy does not implement \SessionHandlerInterface but SessionHandlerProxy requires a \SessionHandlerInterface

Reply:
fixed